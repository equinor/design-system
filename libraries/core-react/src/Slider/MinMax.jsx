import React from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'
import { typographyTemplate } from '../_common/templates'
import { slider as tokens } from './Slider.tokens'

const { enabled } = tokens

const StyledMinMax = styled.span`
  grid-row: 2;
  ${typographyTemplate(enabled.output.typography)}
  position: absolute;
  left: 2px;
  top: ${enabled.track.realHeight};
  /* Avoid track hover on minmax mouse over */
  pointer-events: none;
  text-align: left;
  /** Center align the text with the dot */
  transform: translate(calc(-1 * calc((100% - 8px) / 2)));
  &:last-child {
    left: auto;
    right: 2px;
    transform: translate(calc((100% - 8px) / 2));
  }
`

export const MinMax = ({ children }) => {
  return <StyledMinMax>{children}</StyledMinMax>
}

MinMax.propTypes = {
  /** @ignore */
  children: PropTypes.node.isRequired,
}
