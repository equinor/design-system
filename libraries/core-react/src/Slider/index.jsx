// @ts-nocheck
export { Slider } from './Slider'
