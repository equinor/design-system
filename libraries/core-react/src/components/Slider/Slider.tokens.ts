import { tokens } from '@equinor/eds-tokens'
import type { ComponentToken } from '@equinor/eds-tokens'

const {
  colors: {
    ui: {
      background__light: { rgba: backgroundColor },
      background__default: { rgba: backgroundColorDefault },
      background__medium: { rgba: backgroundColorMedium },
    },
    interactive: {
      primary__resting: { rgba: indicatorColor },
      primary__hover_alt: { rgba: primaryHoverAlt },
      primary__hover: { rgba: primaryHover },
      focus: { rgba: focusOutlineColor },
      disabled__fill: { rgba: backgroundColorDisabled },
      disabled__border: { rgba: backgroundBorderDisabled },
    },
    text: {
      static_icons__tertiary: { rgba: textColor },
    },
  },
  typography: { paragraph },
} = tokens

type SliderToken = ComponentToken

type Slider = {
  enabled: {
    background: string
    track: {
      background: string
      height: string
      realHeight: string
      bottomOffset: string
      indicator: {
        color: string
        hover: {
          color: string
        }
      }
      hover: {
        background: string
      }
    }
    output: {
      height: string
      typography: Typography
      text: string
    }
    handle: {
      background: string
      size: string
      border: {
        color: string
        radius: string
        width: string
        type: string
      }
      outline: string
      outlineOffset: string
      hover: {
        background: string
        border: {
          color: string
        }
      }
    }
    dot: {
      size: string
      border: {
        color: string
        width: string
        type: string
        radius: string
      }
    }
  }
  disabled: {
    background: string
    border: {
      color: string
    }
    track: {
      indicator: {
        color: string
      }
    }
  }
}

export const sliderOld: Slider = {
  enabled: {
    background: backgroundColorDefault,
    track: {
      background: backgroundColor,
      height: '4px',
      realHeight: '24px',
      bottomOffset: '9px',
      indicator: {
        color: indicatorColor,
        hover: {
          color: primaryHover,
        },
      },
      hover: {
        background: backgroundColorMedium,
      },
    },
    output: {
      height: '14px',
      typography: paragraph.overline,
      text: textColor,
    },
    handle: {
      background: backgroundColorDefault,
      size: '12px',
      border: {
        color: indicatorColor,
        radius: '50%',
        width: '2px',
        type: 'solid',
      },
      outline: `1px dashed ${focusOutlineColor}`,
      outlineOffset: '2px',
      hover: {
        background: primaryHoverAlt,
        border: {
          color: primaryHover,
        },
      },
    },
    dot: {
      size: '6px',
      border: {
        color: backgroundColorMedium,
        width: '1px',
        type: 'solid',
        radius: '50%',
      },
    },
  },
  disabled: {
    background: backgroundColorDisabled,
    border: {
      color: backgroundColorMedium,
    },
    track: {
      indicator: {
        color: backgroundBorderDisabled,
      },
    },
  },
}

export const slider: SliderToken = {
  background: backgroundColorDefault,
  entities: {
    track: {
      background: backgroundColor,
      height: '4px',
      spacings: {
        top: '24px',
        bottom: '9px',
      },
      entities: {
        indicator: {
          background: indicatorColor,
          states: {
            hover: {
              background: primaryHover,
            },
          },
        },
      },
      states: {
        hover: {
          background: backgroundColorMedium,
        },
        disabled: {
          background: backgroundBorderDisabled,
        },
      },
    },
    output: {
      height: '14px',
      typography: {
        ...paragraph.overline,
        color: textColor,
      },
    },
    handle: {
      background: backgroundColorDefault,
      height: '12px',
      width: '12px',
      border: {
        type: 'border',
        color: indicatorColor,
        radius: '50%',
        width: '2px',
        style: 'solid',
      },
      states: {
        focus: {
          outline: {
            type: 'outline',
            color: focusOutlineColor,
            width: '1px',
            style: 'dashed',
            offset: '2px',
          },
        },
        hover: {
          background: primaryHoverAlt,
          border: {
            type: 'border',
            color: primaryHover,
          },
        },
      },
    },
    dot: {
      height: '6px',
      width: '6px',
      border: {
        type: 'border',
        color: backgroundColorMedium,
        width: '1px',
        style: 'solid',
        radius: '50%',
      },
      spacings: {
        bottom: '8px',
      },
    },
  },
  states: {
    disabled: {
      background: backgroundColorDisabled,
      border: {
        color: backgroundColorMedium,
      },
    },
  },
}
