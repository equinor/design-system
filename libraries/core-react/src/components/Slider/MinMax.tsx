import * as React from 'react'
import { forwardRef, HTMLAttributes, ReactNode } from 'react'
import styled from 'styled-components'
import { typographyTemplate } from '../../utils'
import { slider as tokens } from './Slider.tokens'

const { enabled } = tokens

const StyledMinMax = styled.span`
  grid-row: 2;
  ${typographyTemplate(enabled.output.typography)}
  position: absolute;
  left: 2px;
  top: ${enabled.track.realHeight};
  /* Avoid track hover on minmax mouse over */
  pointer-events: none;
  text-align: left;
  /** Center align the text with the dot */
  transform: translate(calc(-1 * calc((100% - 8px) / 2)));
  &:last-child {
    left: auto;
    right: 2px;
    transform: translate(calc((100% - 8px) / 2));
  }
`

type MinMaxProps = {
  /** Children is required */
  children: ReactNode
} & HTMLAttributes<HTMLSpanElement>

export const MinMax = forwardRef<HTMLSpanElement, MinMaxProps>(function MinMax(
  { children },
  ref,
) {
  return <StyledMinMax ref={ref}>{children}</StyledMinMax>
})
