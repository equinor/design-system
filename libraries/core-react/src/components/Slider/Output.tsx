import * as React from 'react'
import { forwardRef, OutputHTMLAttributes, ReactNode } from 'react'
import styled from 'styled-components'
import { typographyTemplate } from '../_common/templates'
import { slider as tokens } from './Slider.tokens'

const { enabled } = tokens

type StyledProps = Pick<OutputProps, 'value'>

const StyledOutput = styled.output<StyledProps>`
  --val: ${({ value }) => value};
  --realWidth: calc(100% - 12px);
  width: fit-content;
  position: relative;
  z-index: 1;
  color: ${enabled.output.text};
  ${typographyTemplate(enabled.output.typography)}
  background: ${enabled.background};
  padding: 0 5px;
  top: ${enabled.track.realHeight};
  pointer-events: none;
  /* Calculate the distance on the track*/
  margin-left: calc((var(--val) - var(--min)) / var(--dif) * var(--realWidth));
  /* Idea: Transform negative ((width of outline elem - handle width) / 2 (half of width for centering)) */
 transform: translate(calc(-1 * calc(var(--realWidth) / 2)));
  grid-row: 2;
  grid-column: 1 / -1;
`

type OutputProps = {
  /** Value */
  value: number
  /** HtmlFor */
  htmlFor: string
  /** Children are required */
  children: ReactNode
} & OutputHTMLAttributes<HTMLOutputElement>

export const Output = forwardRef<HTMLOutputElement, OutputProps>(
  function Output({ children, value, htmlFor }, ref) {
    return (
      <StyledOutput ref={ref} value={value} htmlFor={htmlFor}>
        {children}
      </StyledOutput>
    )
  },
)
