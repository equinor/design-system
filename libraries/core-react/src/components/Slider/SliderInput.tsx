import { forwardRef, InputHTMLAttributes } from 'react'
import styled, { css } from 'styled-components'
import { slider as tokens } from './Slider.tokens'
import { bordersTemplate, outlineTemplate } from '../../utils'

const {
  entities: { handle },
  states: { disabled: _disabled },
} = tokens

const track = css`
  width: 100%;
  height: 100%;
  cursor: pointer;
  background: none;
`

const thumb = css`
  ${bordersTemplate(handle.border)}
  height: ${handle.height};
  width: ${handle.width};
  background: ${handle.background};
  cursor: pointer;
  position: relative;
  z-index: 1;
`

const thumbHover = css`
  box-shadow: 0px 0px 0px 6px ${handle.states.hover.background};
  border-color: ${handle.states.hover.border.type === 'border' &&
  handle.states.hover.border.color};
`

const thumbHoverAndDisabled = css`
  cursor: not-allowed;
  box-shadow: none;
`

const thumbDisabled = css`
  background-color: ${_disabled.background};
  border-color: ${_disabled.border.type === 'border' && _disabled.border.color};
`

type SliderInput = {
  type: string
}

const StyledSliderInput = styled.input.attrs<SliderInput>(() => ({
  type: 'range',
}))`
  &::-webkit-slider-runnable-track,
  &::-webkit-slider-thumb,
  & {
    -webkit-appearance: none;
  }
  ::-moz-focus-outer {
    border: 0;
  }

  width: 100%;
  background: transparent;
  background: none;
  grid-column: 1 / -1;
  grid-row: 2;
  font: inherit;
  margin: 0;
  z-index: 2;
  outline: none;
  &[data-focus-visible-added]:focus {
    z-index: 2;
    &::-webkit-slider-thumb {
      ${outlineTemplate(handle.states.focus.outline)};
    }
    &::-moz-range-thumb {
      ${outlineTemplate(handle.states.focus.outline)};
    }
  }
  &:hover,
  &:active {
    &::-webkit-slider-thumb {
      ${thumbHover}
    }
    &::-moz-range-thumb {
      ${thumbHover}
    }
  }
  &:disabled {
    &::-webkit-slider-thumb {
      ${thumbDisabled}
    }
    &::-moz-range-thumb {
      ${thumbDisabled}
    }
  }
  &:disabled:hover,
  &:disabled:active {
    &::-webkit-slider-thumb {
      ${thumbHoverAndDisabled}
    }
    &::-moz-range-thumb {
      ${thumbHoverAndDisabled}
    }
    &::-webkit-slider-runnable-track {
      cursor: not-allowed;
    }
    &::-moz-range-track {
      cursor: not-allowed;
    }
  }

  &:before,
  &:after {
  }
  &:after {
    right: 0;
  }

  /* Must be seperated code blocks for webkit and moz otherwise nothing will be applied */
  &::-webkit-slider-thumb {
    ${thumb}
    margin-top: 6px;
  }
  &::-moz-range-thumb {
    ${thumb}
    /* Avoid too small circles, dunno why this is happening :/  */
    height: 8px;
    width: 8px;
  }
  &::-webkit-slider-runnable-track {
    ${track}
  }
  &::-moz-range-track {
    ${track}
  }
`

type SliderInputProps = {
  /** Slider value */
  value: number
  /** Change function , this is a controlled component */
  onChange: (event: MouseEvent | KeyboardEvent) => void
  /**  Mouse up handler */
  onMouseUp: (event: MouseEvent) => void
  /** Key up handler */
  onKeyUp: (event: KeyboardEvent) => void
  /** Min value */
  min: number
  /** Max value */
  max: number
  /** Id for the input element */
  id: string
  /** Step value */
  step: number
  /** Disabled */
  disabled: boolean
} & InputHTMLAttributes<HTMLInputElement>

export const SliderInput = forwardRef<HTMLInputElement, SliderInputProps>(
  function SliderInput(
    {
      value,
      min,
      max,
      id,
      step,
      onChange,
      onMouseUp,
      onKeyUp,
      disabled,
      ...restProps
    },
    ref,
  ) {
    return (
      <StyledSliderInput
        {...restProps}
        value={value}
        ref={ref}
        min={min}
        max={max}
        id={id}
        step={step}
        onChange={(event) => {
          onChange(event)
        }}
        onMouseUp={(event) => {
          onMouseUp(event)
        }}
        onKeyUp={(event) => {
          onKeyUp(event)
        }}
        disabled={disabled}
      />
    )
  },
)

SliderInput.displayName = 'SliderInput'
