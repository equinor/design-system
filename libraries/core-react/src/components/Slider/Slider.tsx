import {
  forwardRef,
  useState,
  useRef,
  HTMLAttributes,
  MouseEvent,
  KeyboardEvent,
  FormEvent,
} from 'react'
import styled, { css } from 'styled-components'
import { slider as tokens } from './Slider.tokens'
import { MinMax } from './MinMax'
import { Output } from './Output'
import { SliderInput } from './SliderInput'
import { bordersTemplate } from '../../utils'

const {
  entities: { track, handle, dot },
} = tokens

const fakeTrackBg = css`
  background-image: url("data:image/svg+xml,<svg xmlns='http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg'><rect x='0' y='11' fill='${track.background}' width='100%' height='4' rx='2' /></svg>");
  background-size: cover;
  background-repeat: no-repeat;
`

const fakeTrackBgHover = css({
  backgroundImage: `url("data:image/svg+xml,<svg xmlns='http%3A%2F%2Fwww.w3.org%2F2000%2Fsvg'><rect x='0' y='11' fill='${track.states.hover.background}' width='100%' height='4' rx='2' /></svg>")`,
})

const trackFill = css`
  grid-column: 1 / span 2;
  grid-row: 2;
  height: ${track.height};
  margin-bottom: ${track.spacings.bottom};
  align-self: end;
  content: '';
`

const wrapperGrid = css`
  display: grid;
  grid-template-rows: max-content 24px;
  grid-template-columns: 1fr 1fr;
  width: 100%;
  position: relative;
`
type RangeWrapperProps = {
  valA: number
  valB: number
} & Pick<SliderProps, 'min' | 'max' | 'disabled'>

const RangeWrapper = styled.div<RangeWrapperProps>`
  --a: ${({ valA }) => valA};
  --b: ${({ valB }) => valB};
  --min: ${({ min }) => min};
  --max: ${({ max }) => max};
  --dif: calc(var(--max) - var(--min));
  --realWidth: calc(100% - 12px);
  ${wrapperGrid}
  ${fakeTrackBg}
  &::before,
  &::after {
    ${trackFill}
    background: ${({ disabled }) =>
      disabled
        ? track.entities.indicator.states.disabled.background
        : track.entities.indicator.background};
  }
  /** Faking the active region of the slider */
  &::before {
    margin-left: calc(
      calc(${handle.width} / 2) + (var(--a) - var(--min)) / var(--dif) *
        var(--realWidth)
    );
    width: calc((var(--b) - var(--a)) / var(--dif) * var(--realWidth));
  }

  &::after {
    margin-left: calc(
      calc(${handle.width} / 2) + (var(--b) - var(--min)) / var(--dif) *
        var(--realWidth)
    );
    width: calc((var(--a) - var(--b)) / var(--dif) * var(--realWidth));
  }
  @media (hover: hover) and (pointer: fine) {
    &:hover:not([disabled]) {
      ${fakeTrackBgHover}
      &::before,
      &::after {
        background: ${track.entities.indicator.states.hover.background};
      }
    }
  }
`

type WrapperProps = Pick<SliderProps, 'min' | 'max' | 'disabled' | 'value'>

const Wrapper = styled.div<WrapperProps>`
  --min: ${({ min }) => min};
  --max: ${({ max }) => max};
  --dif: calc(var(--max) - var(--min));
  --value: ${({ value }) => value};
  --realWidth: calc(100% - 12px);
  ${wrapperGrid}
  ${fakeTrackBg}
  &::after {
    ${trackFill}
    background: ${({ disabled }) =>
      disabled
        ? track.entities.indicator.states.disabled.background
        : track.entities.indicator.background}
  }
  &::after {
    margin-right: calc(
      (var(--max) - var(--value)) / var(--dif) * var(--realWidth)
    );
    /* Adjusting for start dot circle */
    margin-left: 3px;
  }
  @media (hover: hover) and (pointer: fine) {
    &:hover:not([disabled]) {
      ${fakeTrackBgHover}
      &::after {
        background: ${track.entities.indicator.states.hover.background};
      }
    }
  }
`

const WrapperGroupLabel = styled.div`
  grid-row: 1;
  grid-column: 1 / 3;
`
const WrapperGroupLabelDots = styled(WrapperGroupLabel)`
  &:before,
  &:after {
    content: ' ';
    display: block;
    position: absolute;
    width: ${dot.width};
    height: ${dot.height};
    background: ${tokens.background};
    ${bordersTemplate(dot.border)}
    bottom: ${dot.spacings.bottom};
    left: 0;
  }
  &:after {
    right: 0;
    left: auto;
  }
`

const SrOnlyLabel = styled.label`
  position: absolute;
  width: 1px;
  height: 1px;
  padding: 0;
  margin: -1px;
  overflow: hidden;
  clip: rect(0, 0, 0, 0);
  white-space: nowrap;
  border-width: 0;
`

export type SliderProps = {
  /** Id for the elements that labels this slider */
  ariaLabelledby: string
  /** Components value, range of numbers */
  value: number[] | number
  /** Function to be called when value change */
  onChange?: (
    event: FormEvent<HTMLDivElement>,
    newValue: number[] | number,
  ) => void
  /** Function to be called when value is committed by mouseup event */
  onChangeCommitted?: (
    event: MouseEvent | KeyboardEvent,
    newValue: number[] | number,
  ) => void
  /** Function for formatting the output, e.g. with dates */
  outputFunction?: (text: number) => string | number
  /** Max value */
  max?: number
  /**  Min value */
  min?: number
  /** Stepping interval */
  step?: number
  /** Show the min and max dots or not */
  minMaxDots?: boolean
  /** Show the min and max values or not */
  minMaxValues?: boolean
  /** Disabled */
  disabled?: boolean
} & Omit<HTMLAttributes<HTMLDivElement>, 'onChange'>

export const Slider = forwardRef<HTMLDivElement, SliderProps>(function Slider(
  {
    min = 0,
    max = 100,
    value = [40, 60],
    outputFunction,
    onChange,
    onChangeCommitted,
    minMaxDots = true,
    minMaxValues = true,
    step = 1,
    disabled,
    ariaLabelledby,
    ...rest
  },
  ref,
) {
  const isRangeSlider = Array.isArray(value)
  const [sliderValue, setSliderValue] = useState(
    isRangeSlider ? value : [value],
  )
  const minRange = useRef<HTMLInputElement>(null)
  const maxRange = useRef<HTMLInputElement>(null)
  const onValueChange = (
    event: FormEvent<HTMLDivElement>,
    valueArrIdx?: number,
  ) => {
    const target = event.target as HTMLInputElement
    const changedValue = parseInt(target.value, 10)
    if (isRangeSlider) {
      const newValue = (sliderValue as number[]).slice()
      newValue[valueArrIdx] = changedValue
      setSliderValue(newValue)
      if (onChange) {
        // Callback for provided onChange func
        onChange(event, newValue)
      }
      return
    }

    setSliderValue([changedValue])
    if (onChange) {
      // Callback for provided onChange func
      onChange(event, [changedValue])
    }
  }
  const handleKeyUp = (event: KeyboardEvent) => {
    if (event.keyCode === 37 || event.keyCode === 39) {
      handleCommitedValue(event)
    }
  }

  const handleCommitedValue = (event: MouseEvent | KeyboardEvent) => {
    if (onChangeCommitted) {
      onChangeCommitted(event, sliderValue as number[])
    }
  }

  const getFormattedText = (text: number) => {
    return outputFunction ? outputFunction(text) : text
  }

  const findClosestRange = (event: MouseEvent) => {
    const target = event.target as HTMLOutputElement | HTMLInputElement
    if (target.type === 'output') {
      return
    }
    const bounds = target.getBoundingClientRect()
    const x = event.clientX - bounds.left
    const inputWidth = minRange.current.offsetWidth
    const minValue = minRange.current.value
    const maxValue = maxRange.current.value
    const diff = max - min

    const normX = (x / inputWidth) * diff + min

    const maxX = Math.abs(normX - parseInt(maxValue))
    const minX = Math.abs(normX - parseInt(minValue))
    if (minX > maxX) {
      minRange.current.style.zIndex = '10'
      maxRange.current.style.zIndex = '20'
    } else {
      minRange.current.style.zIndex = '20'
      maxRange.current.style.zIndex = '10'
    }
  }

  const inputIdA = `${ariaLabelledby}-thumb-a`
  const inputIdB = `${ariaLabelledby}-thumb-b`
  const inputId = `${ariaLabelledby}-thumb`

  return (
    <>
      {isRangeSlider ? (
        <RangeWrapper
          {...rest}
          ref={ref}
          role="group"
          aria-labelledby={ariaLabelledby}
          valA={sliderValue[0] as number}
          valB={sliderValue[1] as number}
          max={max}
          min={min}
          disabled={disabled}
          onMouseMove={findClosestRange}
        >
          {minMaxDots && <WrapperGroupLabelDots />}
          <SrOnlyLabel htmlFor={inputIdA}>Value A</SrOnlyLabel>
          <SliderInput
            type="range"
            ref={minRange}
            value={sliderValue[0] as number}
            max={max}
            min={min}
            id={inputIdA}
            step={step}
            onChange={(event) => {
              onValueChange(event, 0)
            }}
            onMouseUp={(event) => handleCommitedValue(event)}
            onKeyUp={(event) => handleKeyUp(event)}
            disabled={disabled}
          />
          <Output htmlFor={inputIdA} value={sliderValue[0] as number}>
            {getFormattedText(sliderValue[0])}
          </Output>
          {minMaxValues && <MinMax>{getFormattedText(min)}</MinMax>}
          <SrOnlyLabel htmlFor={inputIdB}>Value B</SrOnlyLabel>
          <SliderInput
            type="range"
            value={sliderValue[1] as number}
            min={min}
            max={max}
            id={inputIdB}
            step={step}
            ref={maxRange}
            onChange={(event) => {
              onValueChange(event, 1)
            }}
            onMouseUp={(event) => handleCommitedValue(event)}
            onKeyUp={(event) => handleKeyUp(event)}
            disabled={disabled}
          />
          <Output htmlFor={inputIdB} value={sliderValue[1] as number}>
            {getFormattedText(sliderValue[1])}
          </Output>
          {minMaxValues && <MinMax>{getFormattedText(max)}</MinMax>}
        </RangeWrapper>
      ) : (
        <Wrapper
          {...rest}
          ref={ref}
          max={max}
          min={min}
          value={sliderValue[0] as number}
          disabled={disabled}
        >
          <SliderInput
            type="range"
            value={sliderValue[0] as number}
            min={min}
            max={max}
            step={step}
            id={inputId}
            onChange={(event) => {
              onValueChange(event)
            }}
            disabled={disabled}
            aria-labelledby={ariaLabelledby}
            onMouseUp={(event) => handleCommitedValue(event)}
            onKeyUp={(event) => handleKeyUp(event)}
          />
          <Output htmlFor={inputId} value={sliderValue[0] as number}>
            {getFormattedText(sliderValue[0])}
          </Output>
          {/*  Need an element for pseudo elems :/ */}
          {minMaxDots && <WrapperGroupLabelDots />}
          {minMaxValues && (
            <>
              <MinMax>{getFormattedText(min)}</MinMax>
              <MinMax>{getFormattedText(max)}</MinMax>
            </>
          )}
        </Wrapper>
      )}
    </>
  )
})
