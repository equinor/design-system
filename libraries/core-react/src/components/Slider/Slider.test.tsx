/* eslint-disable no-undef */
import { render, cleanup, fireEvent, screen } from '@testing-library/react'
import '@testing-library/jest-dom'
import { Slider } from './Slider'
import type { SliderProps } from './Slider'

afterEach(cleanup)

const getUnixTime = (iso) => {
  return new Date(iso).getTime()
}

const DateSlider = ({
  value,
  ariaLabelledby = 'date-range-slider',
}: SliderProps) => {
  function outputFunction(val) {
    const date = new Date(parseInt(val, 10))
    // The test node server doesn't have full i18n capabilities, using english is the easiest
    return date.toLocaleDateString('en-US', {
      weekday: 'long',
      year: 'numeric',
      month: 'long',
      day: 'numeric',
    })
  }

  return (
    <Slider
      min={getUnixTime('2020-01-01')}
      max={getUnixTime('2020-01-31')}
      ariaLabelledby={ariaLabelledby}
      step={60 * 60 * 24 * 1000}
      value={value}
      outputFunction={outputFunction}
    />
  )
}

describe('Simple slider', () => {
  it('Matches snapshot', () => {
    const { asFragment } = render(
      <Slider value={0} ariaLabelledby="test-one" />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it('Creates a simple slider when providing a number as value', () => {
    const { container } = render(<Slider value={0} ariaLabelledby="test-one" />)
    const input = container.querySelector('input')

    expect(input).toBeDefined()
  })
  it('Sets the aria label', () => {
    const { container } = render(
      <Slider value={5} ariaLabelledby="test-label" />,
    )
    const input = container.querySelector('input')
    expect(input).toHaveAttribute('aria-labelledby', 'test-label')
  })
  it('Can set min, max and step values', () => {
    const { container } = render(
      <Slider
        value={5}
        min={4}
        max={10}
        step={2}
        ariaLabelledby="test-minmax"
      />,
    )
    const input = container.querySelector('input')
    expect(input).toHaveAttribute('min', '4')
    expect(input).toHaveAttribute('max', '10')
    expect(input).toHaveAttribute('step', '2')
  })
  it('Updates output according to value change', () => {
    const handleChange = jest.fn()
    const { container } = render(
      <Slider value={5} ariaLabelledby="test-value" onChange={handleChange} />,
    )
    const input = container.querySelector('input')
    const outputValue = container.querySelector('output')
    expect(outputValue).toHaveTextContent('5')
    fireEvent.change(input, { target: { value: '6' } })
    expect(outputValue).toHaveTextContent('6')
    expect(handleChange).toHaveBeenCalledTimes(1)
  })
  it('Can format the output if a formatting function is provided', () => {
    const { container } = render(
      <DateSlider
        value={getUnixTime('2020-01-01')}
        ariaLabelledby="date-test"
      />,
    )
    const outputValue = container.querySelector('output')
    const input = container.querySelector('input')
    expect(outputValue).toHaveTextContent('Wednesday, January 1, 2020')
    expect(outputValue).not.toHaveTextContent('getUnixTime')
    expect(input).toHaveValue(getUnixTime('2020-01-01').toString())
  })
  it('Has minimum and maximum values as default', () => {
    const { container } = render(
      <Slider value={5} ariaLabelledby="test-default" min={0} max={100} />,
    )
    expect(container).toHaveTextContent('0')
    expect(container).toHaveTextContent('100')
  })
  it('It is possible to turn off minimum and maximum values', () => {
    const { container } = render(
      <Slider
        value={5}
        ariaLabelledby="test-default-off"
        min={0}
        max={100}
        minMaxValues={false}
      />,
    )
    expect(container).not.toHaveTextContent('0')
    expect(container).not.toHaveTextContent('100')
  })
})

describe('Range slider', () => {
  it('Creates a range slider when providing an array as value', () => {
    const { container } = render(
      <Slider value={[0, 20]} ariaLabelledby="test-array" />,
    )
    const inputs = container.querySelectorAll('input')
    expect(inputs).toHaveLength(2)
  })
  it('Can set min, max and step values', () => {
    const { container } = render(
      <Slider
        value={[5, 6]}
        min={4}
        max={10}
        step={2}
        ariaLabelledby="test-range-minmax"
      />,
    )
    const input = container.querySelector('input')
    expect(input).toHaveAttribute('min', '4')
    expect(input).toHaveAttribute('max', '10')
    expect(input).toHaveAttribute('step', '2')
  })
  it('Updates output according to value change', () => {
    const handleChange = jest.fn()
    const ariaId = 'test-rangechange'
    const { container } = render(
      <Slider value={[3, 6]} ariaLabelledby={ariaId} onChange={handleChange} />,
    )

    const inputA = container.querySelector(`#${ariaId}-thumb-a`)
    const inputB = container.querySelector(`#${ariaId}-thumb-b`)
    const outputA = container.querySelector(`output[for="${ariaId}-thumb-a"]`)
    const outputB = container.querySelector(`output[for="${ariaId}-thumb-b"]`)
    expect(outputA).toHaveTextContent('3')
    expect(outputB).toHaveTextContent('6')
    fireEvent.change(inputA, { target: { value: '4' } })
    expect(outputA).toHaveTextContent('4')
    expect(handleChange).toHaveBeenCalledTimes(1)
    fireEvent.change(inputB, { target: { value: '7' } })
    expect(outputB).toHaveTextContent('7')
    expect(handleChange).toHaveBeenCalledTimes(2)
  })
  it('Can format the output if a formatting function is provided', () => {
    const ariaId = 'test-output-formatting'
    const { container } = render(
      <DateSlider
        value={[getUnixTime('2020-01-01'), getUnixTime('2020-01-31')]}
        ariaLabelledby={ariaId}
      />,
    )
    const outputA = container.querySelector(`output[for="${ariaId}-thumb-a"]`)
    const outputB = container.querySelector(`output[for="${ariaId}-thumb-b"]`)
    const inputA = container.querySelector(`#${ariaId}-thumb-a`)
    const inputB = container.querySelector(`#${ariaId}-thumb-b`)
    expect(outputA).toHaveTextContent('Wednesday, January 1, 2020')
    expect(outputA).not.toHaveTextContent('getUnixTime')
    expect(outputB).toHaveTextContent('Friday, January 31, 2020')
    expect(outputB).not.toHaveTextContent('getUnixTime')
    expect(inputA).toHaveValue(getUnixTime('2020-01-01').toString())
    expect(inputB).toHaveValue(getUnixTime('2020-01-31').toString())
  })

  it('Sets the aria label', () => {
    render(<Slider value={[2, 4]} ariaLabelledby="test-rangelabel" />)
    const wrapper = screen.getByRole('group')
    expect(wrapper).toHaveAttribute('aria-labelledby', 'test-rangelabel')
  })
  it('Has minimum and maximum values as default', () => {
    const { container } = render(
      <Slider
        value={[2, 4]}
        ariaLabelledby="test-rangedefault"
        min={0}
        max={100}
      />,
    )
    expect(container).toHaveTextContent('0')
    expect(container).toHaveTextContent('100')
  })
  it('It is possible to turn off minimum and maximum values', () => {
    const { container } = render(
      <Slider
        value={[2, 4]}
        ariaLabelledby="test-range-minmaxoff"
        min={0}
        max={100}
        minMaxValues={false}
      />,
    )
    expect(container).not.toHaveTextContent('0')
    expect(container).not.toHaveTextContent('100')
  })
})
