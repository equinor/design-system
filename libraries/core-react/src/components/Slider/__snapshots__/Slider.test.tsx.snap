// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Simple slider Matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-eCssSg imFDhh"
    max="100"
    min="0"
    value="0"
  >
    <input
      aria-labelledby="test-one"
      class="sc-dlfnbm clBnoA"
      id="test-one-thumb"
      max="100"
      min="0"
      step="1"
      type="range"
      value="0"
    />
    <output
      class="sc-gsTCUz enCDSy"
      for="test-one-thumb"
      value="0"
    >
      0
    </output>
    <div
      class="sc-jSgupP sc-gKsewC bxFDfR fzKbTb"
    />
    <span
      class="sc-bdfBwQ hDUCP"
    >
      0
    </span>
    <span
      class="sc-bdfBwQ hDUCP"
    >
      100
    </span>
  </div>
</DocumentFragment>
`;
